core/libp2p/protocol_muxer/multiselect/message_manager.cpp

```diff
@@ -0,0 +1,351 @@
+                            e) {
+  using Error = libp2p::protocol_muxer::MessageManager::ParseError;
+  switch (e) {
+    case Error::MSG_IS_TOO_SHORT:
+      return "message size is less, than a minimum one";
+    case Error::VARINT_WAS_EXPECTED:
+      return "cannot find a varint, where it was supposed to be";
+    case Error::MSG_LENGTH_IS_INCORRECT:
+      return "length of the message is not as specified in varints inside";
+    case Error::PEER_ID_WAS_EXPECTED:
+      return "cannot find a peer id, where it was supposed to be";
+    case Error::NO_P2P_PREFIX:
+      return "cannot find /p2p prefix, where it was supposed to be";
+    case Error::WRONG_PEER_ID:
+      return "peer id is either ill-formed or cannot be found";
+  }
+  return "uknown error";
+}
+
+namespace {
+  using kagome::common::Buffer;
+  using libp2p::multi::Multihash;
+  using libp2p::multi::UVarint;
+  using MultiselectMessage =
+      libp2p::protocol_muxer::MessageManager::MultiselectMessage;
+
+  /// produces protocol string to be sent: '/p2p/PEER_ID/PROTOCOL\n'
+  boost::format kProtocolTemplate = boost::format("/p2p/%1%2\n");
+
+  /// header of Multiselect protocol
+  constexpr std::string_view kMultiselectHeaderString =
+      "/multistream-select/0.3.0";
+
+  /**
+   * Retrieve a varint from a bytes buffer
+   * @param buffer to be seeked
+   * @param pos - position, from which the retrieval should start; after the
+   * function execution it is set to the position AFTER the found varint
+   * @return varint, if it was retrieved; none otherwise
+   */
+  std::optional<UVarint> getVarint(const Buffer &buffer, size_t &pos) {
+    std::vector<uint8_t> varint_candidate;
+    uint8_t current_byte;
+    while (pos < buffer.size()) {
+      current_byte = buffer[pos];
+      varint_candidate.push_back(current_byte);
+      ++pos;
+      // last varint in the array has a 0 most significant bit
+      if ((current_byte << 7) == 0) {
+        return UVarint{varint_candidate};
+      }
+    }
+    return {};
+  }
+
+  /**
+   * Check that a string begins with a specified substring
+   * @param string to be checked
+   * @param substring to be found
+   * @return true, if the string begins with a substring, false otherwise
+   */
+  bool stringBeginsWith(std::string_view string, std::string_view substring) {
+    if (string.size() < substring.size()) {
+      return false;
+    }
+    for (size_t i = 0; i < substring.size(); ++i) {
+      if (string[i] != substring[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Retrieve a line from a buffer, starting from the specified position
+   * @param buffer, from which the line is to be retrieved
+   * @param current_position, from which to get the line; after the execution is
+   * set to a position after the line
+   * @return line in case of success, error otherwise
+   */
+  outcome::result<std::string> lineToString(const Buffer &buffer,
```

Comment:
what is the meaning of "line"?

Reply:
Messages have the format:
```
<varint_length_of_line><payload><\n>
<varint_length_of_line><payload><\n>
```
So, line in this case is the stringified payload + \n byte of the line, which must begin from `buffer + current_position`

Reply:
Going to put this into the docs